### ISSUE TITLE
Problem: there is no way to delete an AIP

### ISSUE THREAD

**Issue by sallain** [artefactual-sdps/enduro#1076]
**Is your feature request related to a problem? Please describe.**

Once an AIP has been created, there is no way to delete it through Enduro. It is possible to delete the AIP in Archivematica or the Storage Service, but this means that 1) users have to leave Enduro, and 2) the deletion is not reflected in Enduro.

**Describe the solution you'd like**

While viewing a single AIP in Enduro, a user with sufficient permission should be given the option to delete the AIP through a UI button.

Similar to Archivematica, deletion should be a two-step process:

1. A user clicks a button to request that an AIP be deleted, and is prompted to provide a reason (text field).
2. A user approves or denies the deletion request, and is prompted to provide a reason (perhaps optional for approval but required for denial?)

Once the deletion request is approved, a API call should be sent to the Storage Service to delete the AIP. The Storage Service should be set to auto-approve deletion requests, so that the user doesn't need to repeat the deletion approval there. Enduro should then receive a confirmation that the package has been deleted (NOTE: based on the current SS API, this may require Enduro to poll the Storage Service for the package's status).

Throughout this workflow, all actions - deletion request, deletion approval/denial, deletion confirmation - should be recorded in an audit log that persists indefinitely (or as long as the user needs it to).

See Miro for the [full proposed workflow](https://miro.com/app/board/uXjVMlYiVgs=/?moveToWidget=3458764601125357317&cot=14), including status updates for the package.

**Describe alternatives you've considered**

We've gone through many iterations of this workflow, but this seems to be the simplest flow without compromising auditability of the deletion action.

**Additional context**

![Swiss Federal Archives (SFA) - Proposed DPS deletion workflow](https://github.com/user-attachments/assets/b5e59d01-e41c-44ab-90f4-167eb7cc4709)

Copy of the Miro diagram as of 22 November 2024.

**Comment by sallain** [comment]
As a first iteration of this feature, the team agreed that we would implement the following:

* UI elements for happy path (deletion button, reason entry box, approval button, etc.)
* New status (`Pending`) for AIPs where a deletion request has been made
* API call to Storage Service to delete AIP
* Some way to receive confirmation from the Storage Service that the AIP was deleted
* New status (`Deleted`) for AIPs that have been successfully deleted
* Audit log to capture who/why/when

The first iteration does not need to include:

* Denial button and associated workflow
* Email or UI notifications
* Reports

**Comment by sallain** [comment]
@scollazo has confirmed the following:

* It is possible to auto-approve deletion requests in the Storage Service
* The deletion workflow deletes replicated AIPs

**Comment by jraddaoui** [comment]
Now that #1117 is almost completed, we can start thinking about this deletion workflow. Based on [Dan's wireframes](https://miro.com/app/board/uXjVMlYiVgs=/?moveToWidget=3458764616772021934&cot=14), this is what I was planning to do, going more in detail on the backend implementation:

- AIP page:
  - Load current AIP.
  - Load current Workflows.
  - Check AIP/Workflows status to:
    - Show/hide enable/disable delete button (based on deletion request attribute).
    - Show mid-workflow alert on top (based on deletion review/request attribute?).
    - Show task review alert inside delete AIP workflow expandable (based on deletion review attribute).
- Request (`POST /storage/aips/{uuid}/deletion-request [storage:aips:deletion:request]`):
  - User submits deletion request with reason.
  - API:
    - Checks AIP status and permissions.
    - Updates AIP status.
    - Creates a DeletionRequest record in the database with request details.
    - Creates a Workflow and review Task in the DB, but not in Temporal.
  - Show error if there is a mismatch between the dashboard and API checks or...
  - Reload current AIP and Workflows in the dashboard (until we add websocket updates).
- Review (`POST /storage/aips/{uuid}/deletion-review [storage:aips:deletion:review]`):
  - User submits deletion review with decision.
  - API checks AIP status, user permissions and that they're not the requester.
  - Reject:
    - Updates DeletionRequest with review details.
    - Updates AIP status, Workflow and Task in the DB.
  - Approve:
    - Updates DeletionRequest with review details.
    - Updates Workflow and review Task in the DB.
    - Triggers Temporal workflow.
    - The Temporal workflow adds extra Tasks, completes the Workflow in the DB and updates the AIP status.

---

Some thoughts along the way:

Thinking about using a Temporal workflow once the request is triggered and use signals to indicate the user review, as we do in the AIP review process from the ingest workflow. The only reason I'd do that right now is to keep the one-to-one relation between the representation of a workflow in Enduro's DB and the Temporal workflows, but I don't think that's a strong enough reason to add the load to Temporal when it's not really needed.

Thinking about read/list endpoints in the API for the DeletionRequests, considering the sensitive user information they may have and how we display the user decision points in the dashboard, I think it's enough to use the Workflow details, but we can revisit if we need to expose those details in the future. For example, if we need/want to hide the review option for the requester we may need to add a new endpoint for that check, instead of including the user info in the Workflow details.

Thinking about relating the DeletionRequest to the AIP or the Workflow in the schema. I initially related them to the AIP, same as the workflows. But then I thought it would be good to have a direct relation to the Workflow which, based on the above, will be the one used to determine the deletion request status in the AIP/dashboard. So I went with a one-to-one relation to the Workflows, the ones related to the AIPs in a one-to-many relation. But I'm still having second thoughts about this, if we delete the Workflow should we cascade to delete the DeletionRequest in the DB? Set it NULL? We'd loose the relation to the AIP. Maybe we should relate them to both?

Thinking about the API endpoints and attributes, I avoided the use of the `DELETE` since we are not going to delete the resource, just mark it as deleted. Then I thought about using `POST /storage/aips/{uuid}/delete` for the request and `PUT /storage/aips/{uuid}/delete` for the review, but I think what I outlined above is clearer.

---

I've created a WIP pull request including the API desing and DB schema changes needed to achieve this implementation:

https://github.com/artefactual-sdps/enduro/pull/1157

**Comment by djjuhasz** [comment]
@jraddaoui The deletion workflow you've described makes sense to me. 

I notice that an approved deletion request doesn't send an AIP deletion request to AMSS — is that something you believe is out of scope for now, or was it an accidental omission?

**Comment by djjuhasz** [comment]
As to the DeletionRequest relation to the AIP or the Workflow:  I don't understand if Workflow refers to a Temporal Workflow, or an entity in the Enduro database.  

I think we should support multiple DeletionRequests for a single AIP (e.g. an initial request may be denied, and then a subsequent request submitted).  A one-to-many relationship between AIP and DeletionRequests could be supported in the database without involving a Workflow entity though.

I would assume the "Delete AIP" Temporal workflow would only be triggered after the DeletionRequest is approved, in which case it doesn't make sense to me to relate the DeletionRequest only to the "Delete AIP" workflow.

**Comment by jraddaoui** [comment]
Thanks @djjuhasz!

> I notice that an approved deletion request doesn't send an AIP deletion request to AMSS — is that something you believe is out of scope for now, or was it an accidental omission?

That would be part of the "The Temporal workflow adds extra Tasks" part, it should be based on the storage location type/source.

> As to the DeletionRequest relation to the AIP or the Workflow: I don't understand if Workflow refers to a Temporal Workflow, or an entity in the Enduro database.

The entity in the DB, the Temporal workflow won't be triggered if the request is rejected. But we'll record it as a Workflow in the DB and show it in the dashboard as other workflows.

> I think we should support multiple DeletionRequests for a single AIP (e.g. an initial request may be denied, and then a subsequent request submitted). A one-to-many relationship between AIP and DeletionRequests could be supported in the database without involving a Workflow entity though.

That was the idea of using a one-to-many relation with AIP, but the Workflows also have a one-to-many with the AIPs and they will be related one-to-one with the DeletionRequests.

> I would assume the "Delete AIP" Temporal workflow would only be triggered after the DeletionRequest is approved, in which case it doesn't make sense to me to relate the DeletionRequest only to the "Delete AIP" workflow.

But a Workflow will be recorded in the database even if the deletion request is rejected and the Temporal workflow is not triggered. In the end, the deletion requests are related to the AIP and they create a Workflow in all cases, so I'll try to keep the relation to both entities.

**Comment by djjuhasz** [comment]
I can't find any examples of what the UI will look like, but I think it's going to be confusing in the code to have a Workflow database entity when there is no corresponding Temporal Workflow.

**Comment by jraddaoui** [comment]
Yeah, it's a bit confusing to have an Enduro entity called Workflow, that maps to a Temporal workflow in most cases, but not in this one if the deletion is rejected...

> Thinking about using a Temporal workflow once the request is triggered and use signals to indicate the user review, as we do in the AIP review process from the ingest workflow. The only reason I'd do that right now is to keep the one-to-one relation between the representation of a workflow in Enduro's DB and the Temporal workflows, but I don't think that's a strong enough reason to add the load to Temporal when it's not really needed.

Also, the change from "Preservation Action" and "Preservation Task" to "Workflow" and "Task" was something that we discussed last week at sync time (https://github.com/artefactual-sdps/enduro/issues/1117#issuecomment-2734562639 and https://github.com/artefactual-sdps/enduro/pull/1154).

**Comment by jraddaoui** [comment]
I had a chat with @sevein talking about the relevance of using a Temporal workflow since the deletion is requested. It ended up being a discussion about the atomicity of the operation, doesn't matter if the Temporal workflow is triggered at request or approval time.

Consider the case where updates are made in Enduro's DB (DeletionRequest, Workflow and Task created), but triggering the Temporal workflow failed. We looked at different options:

- [Transactional outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html)
- [Two-phase commit protocol](https://en.wikipedia.org/wiki/Two-phase_commit_protocol)
- Manual rollback.
- Delegate to Temporal.

The transactional outbox pattern requires a broker and the two-phase commit protocol won't work with Temporal, so we discarded those two options.

The manual rollback would work as outlined above but, it will create/update the entities in a DB transaction and, if the Temporal request fails, it will rollback them manually in another DB transaction. Delegating everything to Temporal will move the DB writes to the Temporal workflow, including the AIP status update, the creation of the DeletionRequest, Workflow and Task.

Moving everything to Temporal implies a bigger delay between the actual user request and the creation of the DB resources and, without websocket updates in Enduro's storage service API, it may take some time and a manual refresh to see the deletion workflow in the dashboard. It also requires to change how we start the Temporal workflow, we'll need to use a deterministic workflow id and reject duplicates in Temporal to avoid concurrency issues. Temporal has a  `WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE_FAILED_ONLY` option that will allow us to run multiple deletion workflows for the same AIP, we just need to end the Temporal workflow as `terminated` or `cancelled` instead of `completed` when the deletion request is rejected by a user.

Ideally, we'd delegate everything to Temporal so we can also take advantage of all its benefits since the beginning of the process, but I'm not sure I'll have the time to do it that way on the first implementation.

**Comment by jraddaoui** [comment]
Initial implementation is "completed" in https://github.com/artefactual-sdps/enduro/pull/1157. In the end I could do it delegating everything to the Temporal workflow and using a signal to pass the user review.

A few considerations about this implementation:

- It should work with MinIO and AMSS location sources.
- ~~Requires to configure the AM pipeline UUID in Enduro.~~
- For AMSS, it requests the deletion and polls for status, approval has to happen in the AMSS (script or manual).
- Approval in Enduro doesn't have a time limit.
- Approval in AMSS needs to happen within two hours after approval in Enduro.
- Rejecting the deletion in Enduro or AMSS will mark the workflow as canceled.
- Any other error (including the AMSS timeout) will mark the workflow as error and keep the AIP as stored in Enduro.
- The Move button has been removed, doesn't work with AMSS locations and it requires a few tweaks to improve integrity now that it can conflict with the deletion workflow.
- The requester cannot review their own requests. However, the review alert and buttons are still shown, even if they won't work. The error in that case is not reported back to the user.
- The review process doesn't have confirmation dialog, didn't have time for it. It also gives us the chance to add a review reason input in there when we add it. It's not part of the wireframes and we discussed it at syn time, but I agree with [David's comment](https://github.com/artefactual-sdps/enduro/pull/1157#discussion_r2019486920) that we should add it.
- There are two new API endpoint for the request/review:
  - `/storage/aips/{uuid}/deletion-request` requires `storage:aips:deletion:request` attribute.
  - `/storage/aips/{uuid}/deletion-review` requires `storage:aips:deletion:review` attribute.
- Both API endpoints require authentication to keep track of the requester and reviewer details.
- The access token must include the `sub` and `iss` claims to guarantee user uniqueness (based on the [OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability)).
- Added AIP pending status, not reflected in the wireframes.
- User details are saved in the DB and their email shown in the task, but there is no audit log for them.
- No notifications.
- No report.

**Comment by jraddaoui** [comment]
Another thing, the deletion request sent to the AMSS from Enduro includes the following required fields with hardcoded values:

- user_id: 123
- user_email: enduro@example.com

**Comment by djjuhasz** [comment]
> Another thing, the deletion request sent to the AMSS from Enduro includes the following required fields with hardcoded values:
> 
>     * user_id: 123
> 
>     * user_email: [enduro@example.com](mailto:enduro@example.com)

@jraddaoui to be clear, I think the AMSS delete AIP endpoint requires values for `user_id` and `user_email` but they aren't actually validated or relevant for authorization/authentication, right?

**Comment by jraddaoui** [comment]
Yeah, that's why I used those instead of worrying about configuring a proper one, it works with them.

**Comment by jraddaoui** [comment]
After the latest changes, when the deletion is approved/rejected in the AMSS, it can take up to one minute to reflect the update in Enduro.

**Comment by fiver-watson** [comment]
Works great! Nice job @jraddaoui !

**Comment by jraddaoui** [comment]
Important note that I forgot to include in the previous comment, both API endpoints to request and review the deletion **require** authentication to keep track of the requester and reviewer details. Also, the access token must include the `sub` and `iss` claims to guarantee user uniqueness (based on the [OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability)).

I don't think it makes sense to have a request/review process for a system without authentication, so we may want to add a different endpoint to allow deletions in that case.
